### PR TITLE
Add hot reload for the watch command (#627)

### PR DESCRIPTION
**Pull request by kMutagene**
This is a WIP PR trying to implement hot reload for the `fsdocs watch` command as discussed in #627.

The server side part is done, and the cycle works when adding the following script to the _template.html files:

```javascript
var wsUri = "ws://localhost:8901/websocket";
function init()
{
    websocket = new WebSocket(wsUri);
    console.log('connected')
    websocket.onclose = function(evt) { onClose(evt) };
}
function onClose(evt)
{
    console.log('closing');
    websocket.close();
    document.location.reload();
}
window.addEventListener("load", init, false);
```

But i am not sure what the best way to handle injecting this code is, as it has to be removed when building the actual docs, or the page will be loked in a constant reload cycle. My idea so far would be, when running the watch command, copy the template files, inject the code, and generate files based on that. This would also need an additional handling of updating these copied template files. I am not familiar enough with the code base yet to see what the best approach here is, and where best to intoduce these changes in the code, so feedback is welcome!

### CODE REVIEW

File: src/FSharp.Formatting.CommandTool/BuildCommand.fs [hunk 1]
```diff
@@ -643,33 +651,43 @@ type CoreBuildOptions(watch) =
...
                             if runGeneratePhase2() then
-                                regenerateSearchIndex()) }))
+                                regenerateSearchIndex())
+                    }
+                    |> Async.RunSynchronously
+                    Serve.signalHotReload <- true
+                )
 
+            // handler for signalling to refresh the page to the client
+            let hotReloadHandler _ = Serve.signalHotReload <- true
```
Comment (dsyme) [inline src/FSharp.Formatting.CommandTool/BuildCommand.fs:685]: Does this `hotReloadHandler ` get used anywhere?

**Comment by dsyme** [comment]
> But i am not sure what the best way to handle injecting this code is, as it has to be removed when building the actual docs, or the page will be loked in a constant reload cycle. My idea so far would be, when running the watch command, copy the template files, inject the code, and generate files based on that. This would also need an additional handling of updating these copied template files. I am not familiar enough with the code base yet to see what the best approach here is, and where best to intoduce these changes in the code, so feedback is welcome!

I'm not sure of the best approach here - perhaps a template substitution that's only activated when building `watch` docs?

File: src/FSharp.Formatting.CommandTool/BuildCommand.fs (hunk 1 above)
Comment (kMutagene) [reply src/FSharp.Formatting.CommandTool/BuildCommand.fs:685]: nope, that is indeed a leftover from a previous test, will remove

**Comment by kMutagene** [comment]
> I'm not sure of the best approach here - perhaps a template substitution that's only activated when building watch docs?

That worked well!

**Comment by dsyme** [comment]
@kMutagene This looks ready?

**Comment by kMutagene** [comment]
@dsyme yeah i was just not sure if i should merge my own PR or wait for a review ;)

**Comment by dsyme** [comment]
> @dsyme yeah i was just not sure if i should merge my own PR or wait for a review ;)

Feel free to ask for a review bia @ to myself and other maintainters.  Then if you don't get one after a while you can self-merge :-)

File: docs/styling.md [hunk 2]
```diff
@@ -104,6 +104,12 @@ If you write a new theme by CSS styling please contribute it back to FSharp.Form
 You can do advanced styling by creating a new template.  Add a file `docs/_template.html`, likely starting
 with the existing default template.
 
+to enable hot reload during development with `fsdocs watch` in a custom `_template.html` file, make sure to add the following line to your `<head>` tag:
```
Comment (dsyme) [inline docs/styling.md:107]: Capital "T" at start of line.

Also there is a table of substitutions in content.md I believe, this should be added there

File: src/FSharp.Formatting.CommandTool/BuildCommand.fs [hunk 3]
```diff
@@ -643,33 +683,40 @@ type CoreBuildOptions(watch) =
...
                         do! Async.Sleep(300)
                         lock monitor (fun () ->
                         docsQueued <- false
                         if runDocContentPhase1() then
                             if runDocContentPhase2() then
                                 regenerateSearchIndex()
-                        ) }) )
+                        )
+                    }
+                    |> Async.RunSynchronously
```
Comment (dsyme) [inline src/FSharp.Formatting.CommandTool/BuildCommand.fs:695]: This has changed from Start to RunSynchronously.  If it's RunSynchronously then no async is needed?

File: src/FSharp.Formatting.CommandTool/BuildCommand.fs [hunk 4]
```diff
@@ -643,33 +683,40 @@ type CoreBuildOptions(watch) =
...
                         do! Async.Sleep(300)
                         lock monitor (fun () ->
                         generateQueued <- false
                         if runGeneratePhase1() then
                             if runGeneratePhase2() then
-                                regenerateSearchIndex()) }))
-
+                                regenerateSearchIndex())
+                    }
+                    |> Async.RunSynchronously
```
Comment (dsyme) [inline src/FSharp.Formatting.CommandTool/BuildCommand.fs:712]: Likewise here....

**Comment by dsyme** [comment]
On Start v. RunSynchronously - was this fixing a bug?

File: src/FSharp.Formatting.CommandTool/BuildCommand.fs (hunk 3 above)
Comment (kMutagene) [reply src/FSharp.Formatting.CommandTool/BuildCommand.fs:695]: You are right, this is unnecessary. I was kind of thrown off by the 'fire and forget' nature of Async.Start (via [this](https://docs.microsoft.com/de-de/dotnet/fsharp/tutorials/asynchronous-and-concurrent-programming/async#asyncstart)) and ditched it without second thought thinking i have to wait until the computation is finished. I missed that the hot reload trigger does not have to happen after the whole block but just after the last function call inside it, so i moved it inside the block now using the original Async.start version again.

**Pull request merged**